pyproject.toml

```diff
@@ -49,12 +49,15 @@ grpcio = "^1.30.0"
 grpcio-health-checking = "^1.30.0"
 h5py = { version = "==3.10.0", optional = true }
 lxml = "^4.9.2"
+nltk = "^3.8.1"
```

Comment:
What size is it?

Reply:
We will have to find the way to get idea about this.

Reply:
`nltk_data` is located in `C:\Users\your_username\AppData\Roaming\nltk_data`

It's size on disk is 3.31 GB.

Reply:
So, shall we make these dependencies optional such that advanced search fails if the user hasn't installed them? PyFluent is one thing but it would nice to have all this working out of the box in the console, but that nltk_data dependency needs some justifying to be included in the installed software. How should we handle this?

@hpohekar @mkundu1 @prmukherj

Reply:
The nltk data is not a dependency, it is corpus files of all supported languages. The data should be downloaded on-demand when user performs semantic search on a particular language. Otherwise, the new dependencies are not very big (progress-bars from pip install output of this branch):
```
Downloading regex-2024.5.10-cp311-cp311-win_amd64.whl.metadata (41 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.0/42.0 kB 675.2 kB/s eta 0:00:00
Downloading scipy-1.13.0-cp311-cp311-win_amd64.whl.metadata (60 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.6/60.6 kB 1.6 MB/s eta 0:00:00
Downloading nltk-3.8.1-py3-none-any.whl (1.5 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 3.4 MB/s eta 0:00:00
Downloading pyspellchecker-0.8.1-py3-none-any.whl (6.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.8/6.8 MB 3.7 MB/s eta 0:00:00
Downloading scikit_learn-1.4.2-cp311-cp311-win_amd64.whl (10.6 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.6/10.6 MB 3.8 MB/s eta 0:00:00
Downloading joblib-1.4.2-py3-none-any.whl (301 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 301.8/301.8 kB 3.7 MB/s eta 0:00:00
Downloading regex-2024.5.10-cp311-cp311-win_amd64.whl (268 kB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 269.0/269.0 kB 4.2 MB/s eta 0:00:00
Downloading scipy-1.13.0-cp311-cp311-win_amd64.whl (46.2 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.2/46.2 MB 3.5 MB/s eta 0:00:00
```

Reply:
I have searched to download language packs on-demand but they are not available separately for each language.

https://www.nltk.org/nltk_data/

Reply:
You can track it down from the error messages. After I've deleted the downloaded data and disabled the automatic download, I see only `wordnet` and `omw-1.4` datasets (~36 MB together) are needed for jap lang search (output is truncated):
```
>>> import ansys.fluent.core as pyfluent        
>>> pyfluent.search("フォント", language="jpn") 
**********************************************************************
  Resource wordnet not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('wordnet')

>>> import nltk
>>> nltk.download('wordnet')
>>> pyfluent.search("フォント", language="jpn")
**********************************************************************
  Resource omw-1.4 not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('omw-1.4')

>>> nltk.download("omw-1.4")
>>> pyfluent.search("フォント", language="jpn")

 The most similar API objects are:

<solver_session>.results.graphics.particle_track["<name>"].color_map.font_name (Parameter)
<solver_session>.preferences.Appearance.Charts.Font (Object)
<solver_session>.results.graphics.particle_track["<name>"].color_map.font_automatic (Parameter)
<solver_session>.results.graphics.particle_track["<name>"].color_map.font_size (Parameter)
<solver_session>.tui.preferences.appearance.charts.font (Object)
<solver_session>.tui.display.set.windows.scale.font_size (Command)
```

Reply:
Same is the case for English.

```
>>> pyfluent.search("read", language="eng") 
Traceback (most recent call last):
  File "D:\Repos\VirtualEnvs\lib\site-packages\nltk\corpus\util.py", line 84, in __load
    root = nltk.data.find(f"{self.subdir}/{zip_name}")
  File "D:\Repos\VirtualEnvs\lib\site-packages\nltk\data.py", line 583, in find
    raise LookupError(resource_not_found)
LookupError:
**********************************************************************
  Resource wordnet not found.
  Please use the NLTK Downloader to obtain the resource:

  >>> import nltk
  >>> nltk.download('wordnet')

  For more information see: https://www.nltk.org/data.html

  Attempted to load corpora/wordnet.zip/wordnet/

  Searched in:
    - 'C:\\Users\\hpohekar/nltk_data'
    - 'D:\\Repos\\VirtualEnvs\\nltk_data'
    - 'D:\\Repos\\VirtualEnvs\\share\\nltk_data'
    - 'D:\\Repos\\VirtualEnvs\\lib\\nltk_data'
    - 'C:\\Users\\hpohekar\\AppData\\Roaming\\nltk_data'
    - 'C:\\nltk_data'
    - 'D:\\nltk_data'
    - 'E:\\nltk_data'
**********************************************************************
```

Reply:
@seanpearsonuk @mkundu1 

As mentioned here https://omwn.org/omw1.html from our documentation link.

![image](https://github.com/ansys/pyfluent/assets/106588300/d8d84103-3133-45ca-8961-fc0fb7445d41)

We need only `wordnet` and `omw-1.4` for semantic search of all supported languages and as @mkundu1 already mentioned that total size id only (~36 MB together).

We will update the code as follows -

```
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')
```

Reply:
Note that, we should still avoid the download operation during import as that is not a good practice. We can download the data when user performs first language-specific search. Best if we can get them is downloaded during installation.

Also note that, it looks for some search paths within the virtual environment. We need to explore that option for Ansys Python packaging, the data should be included in the Ansys Python package, if possible.

Reply:
@mkundu1 

We can download wordnet and omw-1.4 during installation using command line as follows

```
(VirtualEnvs) D:\pyfluent>python -m nltk.downloader wordnet omw-1.4
D:\Installations\Python310\lib\runpy.py:126: RuntimeWarning: 'nltk.downloader' found in sys.modules after import of package 'nltk', but prior to execution of 'nltk.downloader'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\hpohekar\AppData\Roaming\nltk_data...
[nltk_data] Downloading package omw-1.4 to
[nltk_data]     C:\Users\hpohekar\AppData\Roaming\nltk_data...

(VirtualEnvs) D:\pyfluent>
```




Only 2 data pack with ~36 MB size

![image](https://github.com/ansys/pyfluent/assets/106588300/c153ac32-7949-44b5-afb9-0af78bae4504)

Reply:
These files should be automatically downloaded when user executes `pip install ansys-fluent-core`. Running custom scripts during installation was supported in setup.py, need to check if it is possible with pyproject.toml.

Reply:
We can use following option in pyproject.toml to execute a script. 

https://python-poetry.org/docs/pyproject/#scripts

```
[tool.poetry.scripts]
my_package_cli = 'my_package.console:run'
```

Reply:
We have handled all above points. Thank you.